### PR TITLE
Extracting individual transform logic and fix encoding

### PR DESCRIPTION
**Pull request by davilimap**
Extracting transform file creation to the individual transformers

Changing transform file creation to (hopefully) fix encoding issues:
* Saved file encoding
* XML file header

Consolidating parameter names (sourcePath, transformPath, destinationPath)

### CODE REVIEW

File: src/Microsoft.VisualStudio.SlowCheetah/Resources/Resources.resx [hunk 1]
```diff
@@ -125,4 +129,24 @@
...
+    <value>{
+}
+</value>
+  </data>
+  <data name="XmlTransform_ContentInfo" xml:space="preserve">
+    <value>For more information on using transformations see the web.config examples at http://go.microsoft.com/fwlink/?LinkId=214134. </value>
+    <comment>Info on XDT transforms, to be added as a comment in the transform file</comment>
+  </data>
+  <data name="XmlTransform_XdtAttributeName" xml:space="preserve">
+    <value>xmlns:xdt</value>
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Resources/Resources.resx]: Don't bother putting non-localized strings in resx like this. Put them as a `const string` in a static class.

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/JsonTransformer.cs [hunk 2]
```diff
@@ -79,22 +119,21 @@ public bool Transform(string source, string transform, string destination)
             }
         }
 
-        private bool TrySaveToFile(Stream result, string destinationFile)
+        private bool TrySaveToFile(Stream result, string sourceFile, string destinationFile)
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/JsonTransformer.cs]: We probably want to revert this change as you found `WriteAllText` destroys the files encoding. Then switch to using your new `GetEncoding` helper.

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformUtilities.cs [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+            }
+
+            if (bom[0] == 0 && bom[1] == 0 && bom[2] == 0xfe && bom[3] == 0xff)
+            {
+                return Encoding.UTF32;
+            }
+
+            using (StreamReader reader = new StreamReader(stream, true))
+            {
+                return reader.CurrentEncoding;
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformUtilities.cs]: Do we need a `reader.Peek()`? What is the position of the reader here?

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs [hunk 4]
```diff
@@ -12,73 +13,34 @@ namespace Microsoft.VisualStudio.SlowCheetah
...
-        /// <returns>True is the file is XML</returns>
-        public static bool IsXmlFile(string filepath)
-        {
-            if (string.IsNullOrWhiteSpace(filepath))
-            {
-                throw new ArgumentNullException(nameof(filepath));
-            }
-
-            if (!File.Exists(filepath))
+            foreach (ITransformer transformer in TransformerCatalog)
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs]: If you prefer LINQ, this can be changed to a `FirstOrDefault`

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs [hunk 5]
```diff
@@ -89,7 +51,15 @@ public static bool IsJsonFile(string filePath)
         /// <returns>True is the file type is supported</returns>
         public static bool IsSupportedFile(string filepath)
         {
-            return IsJsonFile(filepath) || IsXmlFile(filepath);
+            foreach (ITransformer transformer in TransformerCatalog)
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs]: This can be changed to a `return TransformerCatalog.Any(x => x.IsFileSupported(filePath));`

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs [hunk 6]
```diff
@@ -21,59 +22,121 @@ public XmlTransformer()
...
+            if (!File.Exists(sourcePath))
+            {
+                throw new FileNotFoundException(Resources.Resources.ErrorMessage_SourceFileNotFound, sourcePath);
+            }
+
+            // If the file should be overwritten or if it doesn't exist, we create it
+            if (overwrite || !File.Exists(transformPath))
+            {
+                // First, copy the original file to preserve encoding and header
+                File.Copy(sourcePath, transformPath, true);
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs]: Use the `overwrite` variable. `File.Copy(sourcePath, transformPath, overwrite);`

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs [hunk 7]
```diff
@@ -89,7 +51,15 @@ public static bool IsJsonFile(string filePath)
         /// <returns>True is the file type is supported</returns>
         public static bool IsSupportedFile(string filepath)
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs]: arg validation

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs [hunk 8]
```diff
@@ -12,73 +13,34 @@ namespace Microsoft.VisualStudio.SlowCheetah
...
+
         /// <summary>
         /// Gets the appropriate <see cref="ITransformer"/> for the given transformation
         /// </summary>
         /// <param name="source">Path to the file to be transformed</param>
         /// <param name="logger">Logger to be used in the transformer</param>
-        /// <param name="useSections">Wheter or not to use sections while logging</param>
         /// <returns>The appropriate transformer for the given file</returns>
-        public static ITransformer GetTransformer(string source, ITransformationLogger logger, bool useSections)
+        public static ITransformer GetTransformer(string source, ITransformationLogger logger)
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformerFactory.cs]: arg validation

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformUtilities.cs [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Determines a stream's encoding by analyzing its byte order mark (BOM).
+        /// Defaults to ASCII when detection of the text file's endianness fails.
+        /// </summary>
+        /// <param name="stream">The stream to analyze.</param>
+        /// <returns>The detected encoding.</returns>
+        public static Encoding GetEncoding(Stream stream)
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformUtilities.cs]: arg validation

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformUtilities.cs [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+    /// </summary>
+    public static class TransformUtilities
+    {
+        /// <summary>
+        /// Determines a text file's encoding by analyzing its byte order mark (BOM).
+        /// Defaults to ASCII when detection of the text file's endianness fails.
+        /// </summary>
+        /// <param name="filename">The text file to analyze.</param>
+        /// <returns>The detected encoding.</returns>
+        public static Encoding GetEncoding(string filename)
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/TransformUtilities.cs]: arg validation

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs [hunk 11]
```diff
@@ -21,59 +22,121 @@ public XmlTransformer()
...
+        /// <inheritdoc/>
+        public void SetLogger(ITransformationLogger logger)
         {
             if (logger == null)
             {
                 throw new ArgumentNullException(nameof(logger));
             }
 
-            this.logger = new XmlShimLogger(logger, useSections);
+            this.logger = new XmlShimLogger(logger, false);
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs]: when passing in bools directly, name the variable.

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/ITransformer.cs [hunk 12]
```diff
@@ -9,12 +9,33 @@ namespace Microsoft.VisualStudio.SlowCheetah
     public interface ITransformer
     {
         /// <summary>
+        /// Sets the logger to be used by the transformation
+        /// </summary>
+        /// <param name="logger">The external logger</param>
+        void SetLogger(ITransformationLogger logger);
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/ITransformer.cs]: Change to `WithLogger` and returns a new version of itself with the logger set.

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs [hunk 13]
```diff
@@ -21,59 +22,121 @@ public XmlTransformer()
...
+                throw new FileNotFoundException(Resources.Resources.ErrorMessage_SourceFileNotFound, sourcePath);
+            }
+
+            // If the file should be overwritten or if it doesn't exist, we create it
+            if (overwrite || !File.Exists(transformPath))
+            {
+                // First, copy the original file to preserve encoding and header
+                File.Copy(sourcePath, transformPath, true);
+
+                XmlDocument transformDoc = new XmlDocument();
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs]: Are any of the parts here `IDisposable`?

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs [hunk 14]
```diff
@@ -21,59 +22,121 @@ public XmlTransformer()
...
+            {
+                throw new ArgumentNullException(nameof(filePath));
+            }
+
+            if (!File.Exists(filePath))
+            {
+                throw new FileNotFoundException(Resources.Resources.ErrorMessage_FileNotFound, filePath);
+            }
+
+            bool isXmlFile = true;
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs]: alternatively you can `return true;` after `xmlTextReader.Read()` and `return false;` at the end of the function

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs [hunk 15]
```diff
@@ -13,6 +13,9 @@ namespace Microsoft.VisualStudio.SlowCheetah
     /// </summary>
     public class XmlTransformer : ITransformer
     {
+        private const string XdtAttributeName = "xmlns:xdt";
+        private const string XdtAttributeValue = "http://schemas.microsoft.com/XML-Document-Transform";
```
Comment (jviau) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs]: There is some other resources that can be pulled out, like the Json default transform file. It is just `"{ }"`

File: src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs (hunk 13 above)
Comment (davilimap) [inline src/Microsoft.VisualStudio.SlowCheetah/Transformer/XmlTransformer.cs]: No, sir

**Comment by davilimap** [comment]
@jviau Changes done
